File: webapp/frontend/src/types/popup.ts
// ============================================
// Popup Types
// ============================================

export interface Popup {
  id: string;
  title: string;
  image_url: string;
  cta_type?: "route" | "url";
  cta_route?: string;
  position: "center_left" | "center_right" | "top_left" | "top_right" | "bottom_left" | "bottom_right";
  is_active: boolean;
  start_date?: string;
  end_date?: string;
  created_at: string;
  updated_at: string;
}

export interface CreatePopupRequest {
  title: string;
  image_url: string;
  cta_type?: "route" | "url";
  cta_route?: string;
  position?: "center_left" | "center_right" | "top_left" | "top_right" | "bottom_left" | "bottom_right";
  is_active?: boolean;
  start_date?: string;
  end_date?: string;
}

export interface UpdatePopupRequest {
  title?: string;
  image_url?: string;
  cta_type?: "route" | "url";
  cta_route?: string;
  position?: "center_left" | "center_right" | "top_left" | "top_right" | "bottom_left" | "bottom_right";
  is_active?: boolean;
  start_date?: string;
  end_date?: string;
}

export interface PopupListParams {
  page?: number;
  limit?: number;
  search?: string;
}

export const POPUP_POSITIONS = [
  { value: "center_left", label: "Giữa bên trái" },
  { value: "center_right", label: "Giữa bên phải" },
  { value: "top_left", label: "Góc trên bên trái" },
  { value: "top_right", label: "Góc trên bên phải" },
  { value: "bottom_left", label: "Góc dưới bên trái" },
  { value: "bottom_right", label: "Góc dưới bên phải" },
] as const;

File: webapp/backend/internal/models/popup.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Popup represents a floating draggable popup displayed in the mobile app.
type Popup struct {
	ID        uuid.UUID      `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	Title     string         `gorm:"type:varchar(255);not null" json:"title"`
	ImageURL  string         `gorm:"type:varchar(500);not null" json:"image_url"`
	CtaType   string         `gorm:"type:varchar(20);default:'route'" json:"cta_type,omitempty"`
	CtaRoute  string         `gorm:"type:varchar(255)" json:"cta_route,omitempty"`
	Position  string         `gorm:"type:varchar(50);default:'center'" json:"position"`
	Platform  string         `gorm:"type:varchar(20);not null;default:'all'" json:"platform"`
	IsActive  bool           `gorm:"default:true" json:"is_active"`
	StartDate *time.Time     `gorm:"type:timestamptz" json:"start_date,omitempty"`
	EndDate   *time.Time     `gorm:"type:timestamptz" json:"end_date,omitempty"`
	CreatedAt time.Time      `gorm:"type:timestamptz;not null;default:now()" json:"created_at"`
	UpdatedAt time.Time      `gorm:"type:timestamptz;not null;default:now()" json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"type:timestamptz;index" json:"deleted_at,omitempty"`
}

func (Popup) TableName() string {
	return "popups"
}
